File: frontend/src/lib/api/settlement.ts
// 카드 정산 API — 현금/카드사별 매출 입력, 수수료·입금 예정일 조회 (백엔드 /api/v1/settlement)
// POS 연동이 없는 매장이 매일 쓸 수 있는 최소 입력 경로다. 잔 수는 선택 입력.
import { apiFetch } from './client';

const auth = (token: string) => ({ Authorization: `Bearer ${token}` });

export type CardIssuer = {
  code: string;
  name: string;
  default_lag: number;
  lag: number; // 이 매장 기준 입금 소요 영업일
  color: string;
  selectable: boolean; // false면 화면 칩에 노출하지 않는다 (카드사 미지정 자리)
  customized?: boolean; // 사장님이 기본값에서 직접 고친 카드사
};

export type RevenueTier = {
  code: string;
  label: string;
  credit: number;
  check: number;
};

export type SettlementSettings = {
  // 한 번이라도 설정을 저장한 적이 있는지 — false면 설정 마법사를 먼저 띄운다
  configured?: boolean;
  revenue_tier: string;
  tier_label: string;
  credit_fee_pct: number;
  check_fee_pct: number;
  custom_credit_fee?: number | null;
  custom_check_fee?: number | null;
  lag_customized?: boolean;
  issuers: CardIssuer[];
  tiers: RevenueTier[];
};

// 서버가 응답하지 않을 때 쓰는 폴백. 구간 코드·요율은 반드시 백엔드
// (settlement_service.REVENUE_TIERS)와 같아야 한다 — 코드가 어긋나면 화면에서 고른
// 구간을 저장할 때 백엔드가 "알 수 없는 매출 구간"으로 400을 낸다.
export const DEFAULT_SETTLEMENT_SETTINGS: SettlementSettings = {
  configured: false,
  revenue_tier: 'small',
  tier_label: '영세 (연매출 3억 이하)',
  credit_fee_pct: 0.4,
  check_fee_pct: 0.15,
  tiers: [
    { code: 'small', label: '영세 (연매출 3억 이하)', credit: 0.4, check: 0.15 },
    { code: 'mid1', label: '중소1 (3억~5억)', credit: 1.05, check: 0.85 },
    { code: 'mid2', label: '중소2 (5억~10억)', credit: 1.2, check: 1.0 },
    { code: 'mid3', label: '중소3 (10억~30억)', credit: 1.45, check: 1.25 },
    { code: 'general', label: '일반 (30억 초과)', credit: 2.0, check: 1.5 },
    { code: 'custom', label: '직접 입력 (계약서 기준)', credit: 1.5, check: 1.2 },
  ],
  issuers: [
    { code: 'shinhan', name: '신한카드', default_lag: 2, lag: 2, color: '#1F4E9C', selectable: true },
    { code: 'kb', name: 'KB국민카드', default_lag: 2, lag: 2, color: '#6A5A48', selectable: true },
    { code: 'samsung', name: '삼성카드', default_lag: 2, lag: 2, color: '#1428A0', selectable: true },
    { code: 'hyundai', name: '현대카드', default_lag: 2, lag: 2, color: '#2C2C2C', selectable: true },
    { code: 'lotte', name: '롯데카드', default_lag: 2, lag: 2, color: '#B8272D', selectable: true },
    { code: 'hana', name: '하나카드', default_lag: 2, lag: 2, color: '#00857E', selectable: true },
    { code: 'woori', name: '우리카드', default_lag: 2, lag: 2, color: '#0067AC', selectable: true },
    { code: 'nh', name: 'NH농협카드', default_lag: 2, lag: 2, color: '#0A8A3C', selectable: true },
    { code: 'bc', name: 'BC카드', default_lag: 3, lag: 3, color: '#E8380D', selectable: true },
    { code: 'pay', name: '간편결제(페이)', default_lag: 3, lag: 3, color: '#F4B400', selectable: true },
    { code: 'unknown', name: '카드사 미지정', default_lag: 3, lag: 3, color: '#8C6F56', selectable: false },
  ],
};

// 최근 매출로 추정한 우대수수료율 구간 추천 (설정 마법사 1단계에서 사용)
export type TierSuggestion = {
  available: boolean;
  observed_days: number;
  reason?: string;
  recommended_tier?: string;
  tier_label?: string;
  credit_fee_pct?: number;
  check_fee_pct?: number;
  daily_avg?: number;
  annual_estimate?: number;
  confidence?: 'high' | 'medium' | 'low';
  basis?: string;
  note?: string;
};

// "이 금액이 언제 얼마로 들어오는지" 한 건 미리보기
export type SettlementPreview = CardDetail & {
  sale_date: string;
  deposit_weekday: string;
  calendar_days: number;
  skipped: { date: string; label: string; reason: string }[];
  tier_label: string;
};

export type CardDetail = {
  issuer: string;
  issuer_name: string;
  color: string;
  card_type: 'credit' | 'check';
  amount: number;
  cups: number | null;
  fee_pct: number;
  fee: number;
  net: number;
  lag: number;
  deposit_date: string;
};

export type DaySales = {
  date: string;
  cash: number;
  cash_cups: number | null;
  cards: CardDetail[];
  card_total: number;
  total: number;
  fee_total: number;
  net_total: number;
  cups: number | null;
  avg_price: number | null;
  has_entry: boolean;
};

export type DepositBucket = {
  deposit_date: string;
  weekday: string;
  gross: number;
  fee: number;
  net: number;
  settled: boolean;
  items: {
    issuer: string;
    issuer_name: string;
    color: string;
    card_type: 'credit' | 'check';
    sale_date: string;
    amount: number;
    fee: number;
    net: number;
  }[];
};

export type DepositSchedule = {
  today: string;
  schedule: DepositBucket[];
  next_deposit: DepositBucket | null;
  pending_net: number;
  pending_gross: number;
  pending_fee: number;
  this_week_net: number;
  fee_note: string;
};

export type SettlementSummary = {
  days: number;
  cash_total: number;
  card_total: number;
  total: number;
  fee_total: number;
  cash_ratio: number | null;
  today_total: number;
  last_week_same_day: number;
  last_week_change_pct: number | null;
  issuer_mix: { issuer: string; issuer_name: string; color: string; amount: number; ratio: number }[];
  daily: { date: string; cash: number; card: number; fee: number; total: number }[];
};

export type CardEntryInput = {
  issuer: string;
  amount: number;
  card_type?: 'credit' | 'check';
  cups?: number | null;
};

export const getSettlementSettings = (token: string) =>
  apiFetch<SettlementSettings>('/api/v1/settlement/settings', { headers: auth(token) });

export const updateSettlementSettings = (
  token: string,
  body: {
    revenue_tier?: string;
    custom_credit_fee?: number;
    custom_check_fee?: number;
    lag_overrides?: Record<string, number>;
  },
) =>
  apiFetch<SettlementSettings>('/api/v1/settlement/settings', {
    method: 'PUT',
    headers: auth(token),
    body: JSON.stringify(body),
  });

export const getTierSuggestion = (token: string, lookbackDays = 90) =>
  apiFetch<TierSuggestion>(`/api/v1/settlement/tier-suggestion?lookback_days=${lookbackDays}`, {
    headers: auth(token),
  });

export const getSettlementPreview = (
  token: string,
  params: { amount: number; card_type?: 'credit' | 'check'; issuer?: string },
) => {
  const q = new URLSearchParams({
    amount: String(Math.max(0, Math.round(params.amount))),
    card_type: params.card_type ?? 'credit',
    issuer: params.issuer ?? '',
  });
  return apiFetch<SettlementPreview>(`/api/v1/settlement/preview?${q.toString()}`, {
    headers: auth(token),
  });
};

export const getDaySales = (token: string, date: string) =>
  apiFetch<DaySales>(`/api/v1/settlement/day?date=${date}`, { headers: auth(token) });

export const saveDaySales = (
  token: string,
  body: { date: string; cash: number; cash_cups?: number | null; cards: CardEntryInput[] },
) =>
  apiFetch<DaySales>('/api/v1/settlement/day', {
    method: 'POST',
    headers: auth(token),
    body: JSON.stringify(body),
  });

export const getDeposits = (token: string, aheadDays = 21) =>
  apiFetch<DepositSchedule>(`/api/v1/settlement/deposits?ahead_days=${aheadDays}`, {
    headers: auth(token),
  });

export const getSettlementSummary = (token: string, days = 28) =>
  apiFetch<SettlementSummary>(`/api/v1/settlement/summary?days=${days}`, { headers: auth(token) });

File: backend/app/api/v1/settlement.py
"""카드 정산 API (백엔드 B) — 현금/카드사별 매출 입력과 입금 예정·수수료 조회

POS 연동이 없는 매장을 위한 핵심 입력 경로다. 메뉴별 잔 수 대신 '결제수단별 총액'만
받고, 그 대가로 카드사별 입금 예정일·수수료·실입금액을 계산해 돌려준다.
"""

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.core.auth import get_current_user
from app.models.user import User
from app.services.ai import settlement_service

router = APIRouter(prefix="/settlement", tags=["settlement"])


class CardEntryIn(BaseModel):
    issuer: str = Field(..., description="카드사 코드 (shinhan, kb, ...)")
    amount: int = Field(0, ge=0, description="해당 카드사 매출 총액(원)")
    card_type: str = Field("credit", description="credit | check")
    cups: Optional[int] = Field(None, ge=0, description="선택 — 잔 수를 세는 매장만")
    memo: Optional[str] = None


class DaySaveRequest(BaseModel):
    date: str = Field(..., description="매출 발생일 YYYY-MM-DD")
    cash: int = Field(0, ge=0)
    cash_cups: Optional[int] = Field(None, ge=0)
    cards: list[CardEntryIn] = Field(default_factory=list)
    memo: Optional[str] = None


class SettingsUpdate(BaseModel):
    revenue_tier: Optional[str] = None
    custom_credit_fee: Optional[float] = Field(None, ge=0, le=10)
    custom_check_fee: Optional[float] = Field(None, ge=0, le=10)
    lag_overrides: Optional[dict[str, int]] = None


@router.get("/settings", summary="수수료 구간·카드사별 입금 소요일")
def get_settings_api(current_user: User = Depends(get_current_user)):
    return settlement_service.get_settings(current_user.email)


@router.put("/settings", summary="정산 설정 저장")
def update_settings_api(
    body: SettingsUpdate,
    current_user: User = Depends(get_current_user),
):
    try:
        return settlement_service.update_settings(
            current_user.email,
            revenue_tier=body.revenue_tier,
            custom_credit_fee=body.custom_credit_fee,
            custom_check_fee=body.custom_check_fee,
            lag_overrides=body.lag_overrides,
        )
    except settlement_service.SettlementError as e:
        raise HTTPException(400, str(e))


@router.get("/tier-suggestion", summary="최근 매출로 추정한 우대수수료율 구간 추천")
def suggest_tier_api(
    lookback_days: int = 90,
    current_user: User = Depends(get_current_user),
):
    return settlement_service.suggest_tier(current_user.email, lookback_days=lookback_days)


@router.get("/preview", summary="설정 미리보기 — 이 금액이 언제 얼마로 들어오는지")
def preview_api(
    amount: int = 100_000,
    card_type: str = "credit",
    issuer: str = "",
    sale_date: Optional[str] = None,
    current_user: User = Depends(get_current_user),
):
    try:
        return settlement_service.preview(
            current_user.email,
            amount=amount,
            card_type=card_type,
            issuer=issuer,
            sale_date=sale_date,
        )
    except settlement_service.SettlementError as e:
        raise HTTPException(400, str(e))


@router.get("/day", summary="하루치 매출 입력 내용 + 그날 수수료·입금 예정")
def get_day_api(date: str, current_user: User = Depends(get_current_user)):
    try:
        return settlement_service.get_day(current_user.email, date)
    except settlement_service.SettlementError as e:
        raise HTTPException(400, str(e))


@router.post("/day", status_code=201, summary="하루치 매출 저장(덮어쓰기)")
def save_day_api(
    body: DaySaveRequest,
    current_user: User = Depends(get_current_user),
):
    try:
        result = settlement_service.save_day(
            current_user.email,
            body.date,
            cash=body.cash,
            cash_cups=body.cash_cups,
            cards=[c.model_dump() for c in body.cards],
            memo=body.memo,
        )
    except settlement_service.SettlementError as e:
        raise HTTPException(400, str(e))
    # 매출이 바뀌었으니 예측 캐시를 비운다 — 대시보드가 바로 최신을 본다
    try:
        from app.services.ai import forecast_service

        forecast_service.invalidate_forecast_cache(current_user.email)
    except Exception:  # 캐시 무효화 실패가 저장을 되돌릴 이유는 없다
        pass
    return result


@router.get("/deposits", summary="다가오는 카드 대금 입금 예정")
def deposits_api(
    ahead_days: int = 21,
    lookback_days: int = 21,
    current_user: User = Depends(get_current_user),
):
    return settlement_service.upcoming_deposits(
        current_user.email, lookback_days=lookback_days, ahead_days=ahead_days)


@router.get("/summary", summary="최근 기간 매출 요약 (현금·카드 비중, 전주 동요일 대비)")
def summary_api(days: int = 28, current_user: User = Depends(get_current_user)):
    return settlement_service.period_summary(current_user.email, days=days)
